src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -251,67 +380,76 @@ case class MatrixRead(
     this
   }
 
-  def execute(hc: HailContext): MatrixValue[Annotation, Annotation, Annotation] = {
+  def execute(hc: HailContext): MatrixValue = {
     val metadata = fileMetadata.metadata
     val localValue =
       if (dropSamples)
         fileMetadata.localValue.dropSamples()
       else
         fileMetadata.localValue
 
-    val parquetFile = path + "/rdd.parquet"
-
-    implicit val localKOk = kOk
-
-    val orderedRDD =
+    val rdd =
       if (dropVariants)
-        OrderedRDD.empty[Annotation, Annotation, (Annotation, Iterable[Annotation])](hc.sc)
+        OrderedRDD2.empty(hc.sc, typ.orderedRDD2Type)
       else {
-        val localRowType = typ.rowType
-        val rdd = new ReadRowsRDD(hc.sc, path, typ.rowType, nPartitions)
-        .map { rv =>
-          // FIXME expensive
-          val ur = new UnsafeRow(localRowType, rv.region.copy(), rv.offset)
-
-          val v = ur.get(0)
-          val va = ur.get(1)
-          val gs: Iterable[Annotation] = ur.getAs[IndexedSeq[Annotation]](2)
-
-          (v, (va, gs))
+        var rdd = OrderedRDD2(
+          typ.orderedRDD2Type,
+          OrderedPartitioner2(hc.sc,
+            hc.hadoopConf.readFile(path + "/partitioner.json.gz")(JsonMethods.parse(_))),
+          new ReadRowsRDD(hc.sc, path, typ.rowType, nPartitions))
+        if (dropSamples) {
+          val localRowType = typ.rowType
+          rdd = rdd.mapPartitionsPreservesPartitioning { it =>
+            var rv2b = new RegionValueBuilder(null)
```

Comment:
previously you used the no arg constructor instead of passing `null`, seems like we should be consistent.